### PR TITLE
[HWKMETRICS-381] dynamically manage gc_grace_seconds for all tables

### PR DESCRIPTION
**Pull request by jsanda**

### CODE REVIEW

**Comment by jsanda** [comment]
retest this please

**Comment by jsanda** [comment]
As noted in the Javadocs, the checks and updates only happen on keyspace change events. Table change events are ignored because we don't dynamically add/alter tables at runtime. I am also wondering if there should be something in place to protect against concurrent updates.

**Comment by stefannegrea** [comment]
Can this work be used for updating the replication factor on a keyspace later?

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+ */
+public class GCGraceSecondsManager implements SchemaChangeListener {
+
+    public static int DEFAULT_GC_GRACE_SECONDS = 604800;
+
+    private static Logger logger = Logger.getLogger(GCGraceSecondsManager.class);
+
+    private RxSession session;
+
+    private PreparedStatement resetGCGraceSeconds;
```
Comment (stefannegrea) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: This is not used. It is needed or left behind by mistake?

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+    private PreparedStatement getGCGraceSeconds;
```
Comment (stefannegrea) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: Should this be final?

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+     * until schema updates have finished.
+     */
+    public void maybeUpdateGCGraceSeconds() {
+        logger.info("Checking tables in " + keyspace + " to see if gc_grace_seconds needs to be updated");
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        Map<String, String> replication = session.getCluster().getMetadata().getKeyspace(keyspace).getReplication();
+        String replicationFactor = replication.get("replication_factor");
+        Completable check;
+
+        if (getClusterSize() == 1 || replicationFactor.equals("1")) {
```
Comment (stefannegrea) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: Is the replication factor ever not equal to 1 now?

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+            check = updateAllGCGraceSeconds(0);
+        } else {
+            // Need to call Completable.merge in order for subscriptions to happen correctly. See https://goo.gl/l15CRV
+            check = Completable.merge(configurationService.load("org.hawkular.metrics").map(config -> {
```
Comment (stefannegrea) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: There is a method on the ConfigurationService that returns the value of a specific key. I think that would be better to use here.

File: core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    public void initClass() {
+        String nodeAddresses = System.getProperty("nodes", "127.0.0.1");
+        Cluster cluster = new Cluster.Builder()
+                .addContactPoints(nodeAddresses.split(","))
+                .build();
+        configurationService = new ConfigurationService();
+        configurationService.init(rxSession);
+
+        SchemaService schemaService = new SchemaService();
+        schemaService.run(session, keyspace, Boolean.valueOf(System.getProperty("resetdb", "true")));
```
Comment (stefannegrea) [inline core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java]: Is it a good idea to reset the schema midway through this test set? Any possible side effects?

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java (hunk 1 above)
Comment (jsanda) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: See MetricsServiceLifecyle.initGCGraceSecondsManager().

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java (hunk 3 above)
Comment (jsanda) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: Quite possibly yes. We don't manage Cassandra at all like we did with RHQ. See [this issue](https://github.com/openshift/origin-metrics/issues/213) in origin-metrics.

File: core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java (hunk 5 above)
Comment (jsanda) [inline core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java]: These tests use a separate keyspace.

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java (hunk 1 above)
Comment (stefannegrea) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: I do not see anywhere a reference to `resetGCGraceSeconds`

File: core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java (hunk 5 above)
Comment (stefannegrea) [inline core/metrics-core-service/src/test/java/org/hawkular/metrics/core/util/GCGraceSecondsManagerITest.java]: Thanks for the clarification; that makes sense.

File: core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java (hunk 1 above)
Comment (jsanda) [inline core/metrics-core-service/src/main/java/org/hawkular/metrics/core/util/GCGraceSecondsManager.java]: Ah sorry Misunderstood :) Yeah, I need to remove it.

**Comment by jsanda** [comment]
> Can this work be used for updating the replication factor on a keyspace later?

I don't think so, but is there something in particular you have in mind?

**Comment by jsanda** [comment]
retest this please

**Comment by stefannegrea** [comment]
Thank you @jsanda!

**Pull request merged**